### PR TITLE
xml code cleanup and xml validation

### PR DESCRIPTION
**Pull request by jhakala**
Improvements to xml validation and error messages for broken xml, make the xml code less confusing.

Tested at FNAL, please test more.

### CODE REVIEW

File: src/rcms/fm/app/level1/HCALEventHandler.java [hunk 1]
```diff
@@ -1038,9 +1042,6 @@ protected void publishRunInfoSummary() {
 
     // TODO JHak put in run start time and stop times. This was always broken.
 
-    Hashtable<String, String> globalRenamedParams = new Hashtable<String, String>();
-    globalRenamedParams.put(  "LOCAL_RUN_KEY"  ,                 "RUN_CONFIG_SELECTED"        );
```
Comment (kakwok) [inline src/rcms/fm/app/level1/HCALEventHandler.java]: DQM still uses this parameter for the task. we should either ask David to implement to move to use `DQM_TASK` or `LOCAL_RUNKEY_NAME` or keep these for now

Comment (jhakala) [reply src/rcms/fm/app/level1/HCALEventHandler.java]: Yes, @DryRun  should certainly move to `DQM_TASK`

**Comment by jhakala** [comment]
Retested at P5, needed some minor fixes, but works AOK now.

File: src/rcms/fm/app/level1/HCALlevelOneEventHandler.java [hunk 2]
```diff
@@ -187,7 +191,7 @@ else if (useResetForRecover.equals("true")) {
             RunKeySetting.put(new StringT("snippet")   ,new StringT(nodes.item(i).getAttributes().getNamedItem("snippet"   ).getNodeValue()));
           }
           else{
-            String errMessage="Cannot find attribute snippet in this Runkey"+runkeyName+", check the RunConfig entry in userXML!";
+            String errMessage="Cannot find attribute snippet in this Runkey"+runkeyName+", check the LocalRunkey entry in userXML!";
```
Comment (kakwok) [inline src/rcms/fm/app/level1/HCALlevelOneEventHandler.java]: I think RunConfig here is correct.

Comment (jhakala) [reply src/rcms/fm/app/level1/HCALlevelOneEventHandler.java]: RunConfig is correct inasmuch as it correctly matches what appears in the grandmaster, but I would argue that we should change the name both in the code (here) and in the grandmasters to something other than RunConfig. Here I chose the name "LocalRunKey" -- if we agree that this is a decent name then we should keep this change and also change what appears in the grandmaster, as per #35 and as per the comment added to the code here: https://github.com/HCALRunControl/levelOneHCALFM/compare/xmlCleanupAndValidation#diff-f7747803ec2cfd6e1d9406833ed3f9d4R179

File: src/rcms/fm/app/level1/HCALParameters.java [hunk 3]
```diff
@@ -85,10 +85,10 @@ public synchronized void initializeParameters() throws ParameterException {
...
-		this.put( new FunctionManagerParameter<VectorT<StringT>> ("AVAILABLE_LOCALRUNKEYS"   ,  new VectorT<StringT>()    ) );  // List of local run keys
-		this.put( new FunctionManagerParameter<StringT>  ("RUN_CONFIG_SELECTED"              ,  new StringT("not set")    ) );  // User selected local run type
-		this.put( new FunctionManagerParameter<StringT>  ("CFGSNIPPET_KEY_SELECTED"          ,  new StringT("not set")    ) );  // Key name for the local run type selected by the user (aka Mastersnippet file)
+		this.put( new FunctionManagerParameter<MapT<MapT<StringT>>>  ("LOCAL_RUNKEY_MAP"     ,  new MapT<MapT<StringT>>()));    // Local run types available
+		this.put( new FunctionManagerParameter<VectorT<StringT>> ("AVAILABLE_LOCAL_RUNKEYS"  ,  new VectorT<StringT>()    ) );  // List of local run keys
+		this.put( new FunctionManagerParameter<StringT>  ("MASTERSNIPPET_SELECTED"           ,  new StringT("not set")    ) );  // User selected mastersnippet
+		this.put( new FunctionManagerParameter<StringT>  ("LOCAL_RUNKEY_SELECTED"            ,  new StringT("not set")    ) );  // Key name for the local run type selected by the user (aka Mastersnippet file)
```
Comment (kakwok) [inline src/rcms/fm/app/level1/HCALParameters.java]: suggest
`LOCAL_RUNKEY_MAP` to have comment of `//Local run types available: <LocalRunKey name : <LocalRunKey attribute: Value>`
`AVAILABLE_LOCAL_RUNKEYS` to have comment of `//List of local run keys names, keys of LOCAL_RUNKEY_MAP`
`MASTERSNIPPET_SELECTED` to have comment of `//User selected mastersnippet file name`
`LOCAL_RUNKEY_SELECTED` to have comment of `//Key name for the local run key selected by the user`

File: test/userXML.xsd [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0"?>
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="qualified">
+  <xs:element name="userXML" type="userXMLType"/>
+  <xs:complexType name="userXMLType">
+    <xs:sequence>
+      <xs:element type="xs:string" name="CfgCVSBasePath" maxOccurs="1" minOccurs="0"/>
+      <xs:element type="xs:string" name="MasterSnippetList" maxOccurs="1" minOccurs="0"/>
```
Comment (kakwok) [inline test/userXML.xsd]: Just a reminder to update this line when we change the element name we expect.

**Comment by jhakala** [comment]
OK, this is ready for deployment, provided we make the changes:

in the userXML:
`MasterSnippetList`  -> `grandmaster`

in the grandmaster snippets:
`RunConfig` -> `LocalRunkey`

Please let me know when to deployment is scheduled in case any unforeseen issues arise (however, this has been tested at P5 successfully already but with the old names in the userXML and grandmasters).

**Comment by jhakala** [comment]
Suggested release note for this PR:

If you install HCALFM version >= [this release], you must add the files grandmaster.xsd and userXML.xsd to your CfgCVS. They must appear in the directory /Master at the top level of your CfgCVS. In other words, if your CfgCVSBasePath in your configuration's Level1FM userXML is, for example, `/home/daqowner/CfgCVS`, then the .xsd files must appear in `home/daqowner/CfgCVS/Master/grandmaster.xsd` and `home/daqowner/CfgCVS/Master/userXML.xsd`.

**Pull request merged**